docs/proposal/20220929-ipam-support.md

```diff
@@ -0,0 +1,422 @@
+   configure the IPAddress on the correct device.
+
+1. Update the VSphereMachine status to reflect the network device state. Update
+   the IPAddressClaim condition.
+
+    ```yaml
+    ---
+    status:
+      conditions:
+        - lastTransitionTime: "2022-09-14T14:00:00Z"
+          type: IPAddressClaimed
+          status: "True"
+      network:
+        - connected: true
+          ipAddrs:
+            - 10.10.10.100
+          macAddr: 00:01:02:03:04:05
+          networkName: VM Network
+    ```
+
+#### Owner Reference
+
+When a VSphereMachine is deleted, an existing finalizer prevents the removal of
+the VSphereMachine object until the VM in vSphere is gone. As long as the
+IPAddressClaim has an owner reference to the VSphereMachine it will get safely
+deleted when the VSphereMachine is deleted. The deletion of the IPAddressClaim
+when the VSphereMachine is deleted is safe because the VM is deleted and is no
+longer using the IP.
+
+#### Flavors
+
+An IPAM flavor will be added to support configuring an array of group, kind,
+name, VM Network objects. When the IPAM flavor is specified, the devices in the
+network spec of the Nodes will have DHCP turned off. Each of the device's
+fromPool in the network spec will be populated with one of the above mentioned
+configuration objects. The templating and configurability of flavors may
+restrict the possible ways IPAM can be configured through flavors.
+
+#### Validation
+
+In the NetworkDeviceSpec:
+
+The IPAddrs field is required if DHCP4 or DHCP6 is false. The new fromPools
+field allows for a third option and therefore we need to update validation to
+ensure one of the three options is specified for a network device.
+
+#### Nameservers
+
+Nameservers can come from DHCP if enabled. Therefore when DHCP is disabled and
+the `FromPools` field is set  a user will have to supply their own nameservers
+in the device spec.
+
+### Security Model
+
+- Following the original IPAM integration proposal, the VSphereMachine is
+  required to be in the same namespace as the IPAddressClaim, IPAddress, and
+  IPPool.
+  - For IPAddressClaim and IPAddress this is enforced by the lack of a
+    namespace field on the reference
+  - The fromPools field on the VSphereMachineTemplate/VSphereMachine also
+    lacks a namespace field on the reference.
+
+### Risks and Mitigations
+
+- Unresponsive IPAM providers can prevent successful creation of a cluster
+- Any issues with fulfilling the IPAddressClaim can prevent successful creation of a cluster
+  - This is partially by design i.e we don't want the cluster to be created if
+    it can't fulfill the ip address claim, but this is just added complexity
+    that a user has to be aware of.
+- Wrong IP address allocations (e.g. duplicates) can brick a cluster’s network
+- Incorrectly assigning IP address claims to the wrong device
+- How do we handle a user deleting a claim directly? This could create issues

```

Comment:
It could be protected by a finalizer until the machine is deleted. If users delete it regardless they're on their own.

Reply:
We think this is covered in the section named Owner References. There is an existing finalizer on the VSphereVM. The vm object is not deleted until the vm is actually gone. We've added an owner ref (back to the vm) to the claim, and the claim will be cascade deleted when the vm object is deleted. CAPV controllers will not delete claims directly, and relies on the garbage collector. This seems safe to us, and a finalizer didn't seem necessary. If a user deletes the Claim manually, they'll have a bad day.

Reply:
We've been discussing this more. Because deleting a claim can cause an IP to be assigned to multiple devices, which would be a calamity, we think we should add the finalizer to the claim. We'll update the proposal. The finalizer would ensure the vm is no longer present in the vSphere api.